test/Contract/Buy.purs

```diff
@@ -0,0 +1,417 @@
+  }
+
+type BasicBuyAssertGroup = PostBuyTestData -> Array (Contract () Unit)
+
+buyTestConfig1 :: BuyTestConfig BasicBuyAssertGroup
+buyTestConfig1 =
+  { mintParams: mintParams1
+  , expectedShares:
+      { mpGain: BigInt.fromInt $ 10 * 1000000
+      , sellerGain: BigInt.fromInt $ 80 * 1000000
+      , authorGain: BigInt.fromInt $ 10 * 1000000
+      }
+  , retBehaviour: ToMarketPlace
+  , authorIsSeller: true
+  , assertions: nftToMarketPlaceAssert
+  , testName: "no low shares"
+  , skip: false
+  , only: false
+  , shouldError: false
+  }
+
+buyTestConfig2 :: BuyTestConfig BasicBuyAssertGroup
+buyTestConfig2 = buyTestConfig1
+  { mintParams = mintParams2
+  , expectedShares
+      { mpGain = BigInt.fromInt 0
+      , sellerGain = BigInt.fromInt $ 90 * 1000000
+      , authorGain = BigInt.fromInt $ 10 * 1000000
+      }
+  , testName = "low marketplace share"
+  }
+
+buyTestConfig3 :: BuyTestConfig BasicBuyAssertGroup
+buyTestConfig3 = buyTestConfig1
+  { mintParams = mintParams3
+  , expectedShares
+      { mpGain = BigInt.fromInt $ 10 * 1000000
+      , sellerGain = BigInt.fromInt $ 90 * 1000000
+      , authorGain = BigInt.fromInt 0
+      }
+  , testName = "low author share"
+  }
+
+buyTestConfig4 :: BuyTestConfig BasicBuyAssertGroup
+buyTestConfig4 = buyTestConfig2
+  { mintParams = mintParams4
+  , testName = "low author and marketplace shares"
+  , expectedShares
+      { mpGain = BigInt.fromInt 0
+      , sellerGain = BigInt.fromInt $ 100 * 1000000
+      , authorGain = BigInt.fromInt 0
+      }
+  }
+
+buyTestConfig5 :: BuyTestConfig BasicBuyAssertGroup
+buyTestConfig5 = buyTestConfig1
+  { mintParams = mintParams5
+  , testName = "price too low for min ada requirement"
+  , shouldError = true
+  }
+
+buyTestConfig6 :: BuyTestConfig BasicBuyAssertGroup
+buyTestConfig6 = buyTestConfig1
+  { mintParams = mintParams6
+  , testName = "fractional shares (.5)"
+  , expectedShares
+      { mpGain = BigInt.fromInt 5_000_000
+      , sellerGain = BigInt.fromInt 40_000_005
+      , authorGain = BigInt.fromInt 5_000_000
+      }
+  }
+
+buyTestConfig7 :: BuyTestConfig BasicBuyAssertGroup
+buyTestConfig7 = buyTestConfig6
+  { mintParams = mintParams7
+  , testName = "fractional shares (.1)"
+  , expectedShares
+      { sellerGain = BigInt.fromInt 40_000_001
+      , authorGain = BigInt.fromInt 5_000_000
+      }
+  }
+
+buyTestConfig8 :: BuyTestConfig BasicBuyAssertGroup
+buyTestConfig8 = buyTestConfig6
+  { mintParams = mintParams8
+  , testName = "fractional shares (.9)"
+  , expectedShares
+      { sellerGain = BigInt.fromInt 40_000_009
+      , authorGain = BigInt.fromInt 5_000_000
+      }
+  }
+
+addVariants
+  :: (BuyTestConfig BasicBuyAssertGroup -> BuyTestConfig BasicBuyAssertGroup)
+  -> Array (BuyTestConfig BasicBuyAssertGroup)
+  -> Array (BuyTestConfig BasicBuyAssertGroup)
+addVariants vary = Array.uncons >>> case _ of
+  Nothing -> []
+  Just { head: conf, tail: confs } ->
+    conf : vary conf : addVariants vary confs
```

Comment:
This looks like it could be done with a fold
perhaps `addVariants vary = foldMap (\x -> [x, vary x])` or using foldr for something a little more efficient